gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,196 @@
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
+    } catch (FileNotFoundException exception) {
+      // If the metadata file doesn't exist, then there is no key store, just return null.
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException, InterruptedException, GeneralSecurityException {
+    Process process = processProvider.createProcess(metadata);
+
+    // Run the command and timeout after 1000 milliseconds.
+    int exitCode = runCertificateProviderCommand(process, 1000);
+    if (exitCode != 0) {
+      throw new IOException("Cert provider command failed with exit code: " + exitCode);
+    }
+
+    // Create mTLS key store with the input certificates from shell command.
+    return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+  }
+
+  @VisibleForTesting
+  static ImmutableList<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+
+    // In the while loop, keep checking if the process is terminated every 100 milliseconds
+    // until timeout is reached or process is terminated. In getKeyStore we set timeout to
+    // 1000 milliseconds, so 100 millisecond is a good number for the sleep.
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
+      remainTime -= System.currentTimeMillis() - startTime;
+
+      try {
+        // exitValue throws IllegalThreadStateException if process has not yet terminated.
+        // Once the process is terminated, exitValue no longer throws exception. Therefore
+        // in the while loop, we use exitValue to check if process is terminated. See
+        // https://docs.oracle.com/javase/7/docs/api/java/lang/Process.html#exitValue()
+        // for more details.
+        return commandProcess.exitValue();
+      } catch (IllegalThreadStateException ignored) {
+      }
+    }
+
+    commandProcess.destroy();
+    throw new IOException("cert provider command timed out");
```

Comment:
process timeout is normally 124 instead of an exception.

Reply:
Cert provider command is a binary provided by Google. It is supposed to run successfully to give us the certificate so we can move forward with mtls. Anything wrong with this command is an exception. This behavior is the same as the apiary client mtls as well.

Reply:
as above a failure here is not an IOException

Reply:
As above this is implementation details.